### DISCUSSION TITLE
Creating accessor functions from strings

### DISCUSSION THREAD

**Discussion by a3ng7n** [capnproto/capnproto#2080] [Q&A]
Hello!

I have two programs, one repeatedly generating unpacked capnp messages, and one consuming said messages. The messages are of the form (from the example page):
```
struct Person {
  name @0 :Text;
  birthdate @3 :Date;

  email @1 :Text;
  phones @2 :List(PhoneNumber);

  struct PhoneNumber {
    number @0 :Text;
    type @1 :Type;

    enum Type {
      mobile @0;
      home @1;
      work @2;
    }
  }
}

struct Date {
  year @0 :Int16;
  month @1 :UInt8;
  day @2 :UInt8;
}
```

Assume the generating program keeps the lengths of all of the lists the same over time. I want to give the consuming program (at start) a string like "Person.phones[1].number" and have it access the "number" property of the 2nd "PhoneNumber" in the "phones" list property of the "Person" object (message?).

I imagine I could write something that parses "phones" into a `getPhones` call, but I remember reading that repeated calls to getting a non-trivial property isn't a good idea (and the idea would be providing many many strings to the consuming program). I was then thinking I could calculate the offset that the string refers to, but I quickly got lost in backing out the encoding algorithm, and my brain started to hurt thinking about segments. Is there a suggested way to do this?

And although the consuming program is in c++, the generating program is in c, and there's a third program that generates the strings in python; so if there's a relatively cross-language friendly way to do this, e.g. generating offsets, I'd really love that.

Let me know what you think! Thank you!

**Comment by kentonv** [comment]
> I remember reading that repeated calls to getting a non-trivial property isn't a good idea

Well, this is just because of two issues, both of which should be easily solvable in this use case:
1. The pointer needs to be validated again every time it is followed. This is slightly more expensive than it looks, so if you have several lines in a row that are all getting the same pointer it makes a lot of sense to store it once in a local variable instead. But for your use case I just wouldn't worry about this.
2. The "traversal limit" counts the total size of the objects pointed to by pointers you have followed, and errors as soon as this size crosses the limit (default 64MB). If you expect to keep the same MessageReader in memory long-term and keep reading fields of it, you need to make sure to increase the traversal limit to something very large. Or, you could create a new MessageReader each time you want to read the message -- as long as the backing bytes are kept in memory this should be cheap.
